tests/h/search/query_test.py

```diff
@@ -141,12 +141,88 @@ def search(self, search):
 
 
 class TestUriFilter(object):
+    @pytest.mark.parametrize("field", ("uri", "url"))
+    def test_filters_by_field(self, search, Annotation, field):
+        Annotation(target_uri="https://foo.com")
+        expected_ids = [Annotation(target_uri="https://bar.com").id]
+
+        result = search.run({field: "https://bar.com"})
+
+        assert sorted(result.annotation_ids) == sorted(expected_ids)
+
+    def test_filters_on_whole_url(self, search, Annotation):
+        Annotation(target_uri="http://bar.com/foo")
+        expected_ids = [Annotation(target_uri="http://bar.com").id,
+                        Annotation(target_uri="http://bar.com/").id]
+
+        result = search.run({"url": "http://bar.com"})
+
+        assert sorted(result.annotation_ids) == sorted(expected_ids)
+
+    def test_filter_matches_buggy_uri(self, search, Annotation):
+        Annotation(target_uri="https://bar.com")
+        expected_ids = [Annotation(target_uri="buggy-uri").id]
+
+        result = search.run({"uri": "buggy-uri"})
+
+        assert sorted(result.annotation_ids) == sorted(expected_ids)
+
+    def test_filters_aliases_http_and_https(self, search, Annotation):
+        expected_ids = [Annotation(target_uri="http://bar.com").id,
+                        Annotation(target_uri="https://bar.com").id]
+
+        result = search.run({"url": "http://bar.com"})
+
+        assert sorted(result.annotation_ids) == sorted(expected_ids)
+
+    def test_filter_identifies_pdfs_marked_as_same_uri(self, search, Annotation, storage):
```

Comment:
Do we have any smartness/auto detection for stuff like this built in? I thought I heard someone mention in SF that we did but I'm not sure where/what it is specifically.

Reply:
What did you mean by smartness/auto detection?

For this specific test case, we're actually testing something more general that PDFs, which is that the filter matches any URIs equivalent to the test URI. URI equivalence can be used for HTTP URIs as well. To make this clearer, I would suggest:
 - Changing the test name
 - Moving the setting of `expand_uri`'s return value (not the rest of the storage fixture) into this function so it is easier to see why a query for one URI matches a bunch of annotations that have different URIs.

Reply:
I thought we had something that tried to grab the unique id of the pdf and therefore if two people were annotating the same pdf at two different urls it would use the unique id as opposed to the url.

Reply:
Ah, yes we do have logic that gets a unique ID from the PDF, but it works as follows:

- The _client_ reads the unique ID from the PDF, and generates a URI of the form `urn:x-pdf:{fingerprint}`
- The _client_ fetches annotations using an API call of the form `/api/search?uri={http URL of PDF}&uri=urn:x-pdf:{fingerprint}`
- h then returns annotations that match _either_ the HTTP URL or the unique ID